### PR TITLE
KSES: Preserve some additional invalid HTML comment syntaxes.

### PR DESCRIPTION
**Pull request by dmsnell**
Trac ticket: Core-61009

## Status

The one test failure appears to be related to deprecating the outdated IE comment syntax. It's not a problem because the test enshrines a mis-parsing of special HTML comments.

This change fixes that mis-parsing and will prevent further corruption.

## Description

When `wp_kses_split` processes a document it attempts to leave HTML comments alone. It makes minor adjustments, but leaves the comments in the document in its output. Unfortunately it only recognizes one kind of HTML comment and rejects many others.

In HTML there are many kinds of invalid markup which, according to the specification, are to be interpreted as an HTML comment. These include, but are not limited to:

 - HTML comments with invalid syntax, `<!-->`, `<!-- --!>`, etc…
 - HTML closing tags whose tag name is invalid `</3>`, `</%happy>`, etc…
 - Things that look like XML CDATA sections, `<![CDATA[…]]>`
 - Things that look like XML Processor Instruction nodes, `<?include "blarg">`

This patch makes a minor adjustment to the algorithm in wp_kses_split to allow two additional kinds of HTML comments:

 - HTML comments with the incorrect closer `--!>`, because this one was a simple and easy change.
 - Closing tags with an invalid tag name, e.g. `</%dolly>`, because these are required to open up explorations in Gutenberg on Bits, a new iteration of dynamic tokens for externally-sourced data, or "Shortcodes 2.0"

These invalid closing tags, which in a browser are interpreted as comments, are one proposal for a placeholder mechanism in the HTML API unlocking HTML templating, a new kind of shortcode, and more. Having these persist in Core is a requirement for exploring and utilizing the new syntax because as long as Core removes them, there's no way to load content from the database and experiment on the full life cycle of potential Bits systems.

On its own, however, this represents a kind of bug fix for Core, making the implementation of wp_kses_split() more closely align with its stated goal of leaving HTML comments as comments. It doesn't attempt to fully fix the mis-parsed comments (because that is a much deeper issue and involves many more questions about existing expectations) but it does propose a couple of hopefully and expectedly minor fixes that hopefully won't break any existing code or projects.

## Discussion

Concerning the removed `kses` test, as far as I can tell there is no support for conditional comments like these after IE9, which I believe we don't support anymore. All modern browsers will interpret the entire span of content as a single comment.

Unfortunately there's a twist to the attack which follow best-practice recommendations for conditional comments, which is to add a short abruptly-closed comment after the opening.

```html
<!--[if gte IE 4]><!-->
<SCRIPT>alert('XSS');</SCRIPT>
<![endif]-->
```

The addition of `<!-->` ends the opening comment and re-reveals the `<SCRIPT>` tag. However, this should not affect the security of WordPress because it will be looking for the `SCRIPT` tags through other code paths after the removal of this restriction.

> Warning: Should verify this in a clear way.

### CODE REVIEW

**Comment by adamziel** [comment]
The test failure is related to IE conditional comments. However, dealing with those shouldn't matter anymore:

> Conditional comments are conditional statements interpreted by Microsoft Internet Explorer versions 5 through 9 in HTML source code

https://www.sitepoint.com/microsoft-drop-ie10-conditional-comments/

File: src/wp-includes/kses.php [hunk 1]
```diff
@@ -1080,14 +1092,39 @@ function _wp_kses_split_callback( $matches ) {
...
 
-	// Allow HTML comments.
+	/*
+	 * Preserve funky comments.
+	 *
+	 * When a closing tag appears with a name that isn't a valid tag name,
+	 * it should be interpreted as an HTML comment. It extends until the
+	 * first `>` character after the initial opening `</`.
+	 */
+	if ( 1 === preg_match( '~</[^a-zA-Z][^>]*>~', $content ) ) {
```
Comment (adamziel) [inline src/wp-includes/kses.php]: So this bales out an returns `$content` if there is a match anywhere in the string. Is this expected? Would this kick in if that's inside of an attribute?

File: src/wp-includes/kses.php [hunk 2]
```diff
@@ -981,7 +982,18 @@ function wp_kses_split( $content, $allowed_html, $allowed_protocols ) {
 	$pass_allowed_html      = $allowed_html;
 	$pass_allowed_protocols = $allowed_protocols;
 
-	return preg_replace_callback( '%(<!--.*?(-->|$))|(<[^>]*(>|$)|>)%', '_wp_kses_split_callback', $content );
+	$token_pattern = <<<REGEX
+~
+	(                         # Detect comments of various flavors before attempting to find tags.
+		(?:<!--.*?(--!?>|$))  # Well-formed HTML comments like `<!-- ... -->` and also with invalid `--!>` closer.
+		|
+		</[^a-zA-Z][^>]*>     # Closing tags with invalid tag names.
```
Comment (adamziel) [inline src/wp-includes/kses.php]: That regexp used to match everything between < and > but now it only matches invalid tag closers. So this change means valid tag names and valid closers are no longer treated by `_wp_kses_split_callback`. At the same time, the goal of this PR is to preserve HTML funky comments. Why, in order to preserve funky comments, is this passing _less_ information to the replacement callback?

Comment (adamziel) [reply src/wp-includes/kses.php]: Aha – I'm wrong. This entire regexp branch is _a new clause_. Tag-like spans are still matched in the next branch. Cool.

**Comment by adamziel** [comment]
Other than my `preg_match` question, this LGTM. I'd just remove the failing test case, conditional comments were only relevant up to IE9 which [WordPress doesn't support anymore](https://make.wordpress.org/core/handbook/best-practices/browser-support/). And even if it did, the new output doesn't seem like an XSS vector. I'd still get inputs from others, cc @westonruter

**Comment by westonruter** [comment]
Confirmed that Chrome parses these as comments:

```html
<!-- comment 1 -->
foo
<!-- comment 2 --!>
bar
</%comment3>
baz
```

image

How common are these other invalid comments?

I presume you'll be adding tests specifically for the changes here?

**Comment by dmsnell** [comment]
> How common are these other invalid comments?

I'm guessing rare, but it might be a fun experiment to scan some sites and count them. This patch _only_ addresses a single kind of invalid comment leaving most other constructions alone. This is because I'm trying to be tactical here and leave a small surface area of change.

> I presume you'll be adding tests specifically for the changes here?

Yes absolutely. Before I do that though I want to make sure folks are on board with the change. If we use these "funky comments" for Bits, template placeholders, translation wrappers, they could become quite popular.

In a local branch I was playing with replacing `wp_kses_split2` using the HTML API and it provides some great affordances, but as usual, far more questions to resolve.

For instance, WordPress currently strips away `<:::>`, interpreting it as an invalid tag. It's not though, it's legitimate plaintext _despite being invalid markup_. So now what do we do?

 - Do we attempt to maintain this legacy behavior that corrupts documents that otherwise are fine?
 - Do we preserve this legitimate HTML and allow its _surprising_ form to thread through a stack of code expecting more familiar HTML?
 - Do we eagerly replace confusing HTML like this before passing it along into WordPress, taking a performance hit to provide regularity and mitigate further misinterpretation by later code on the render chain?

There are further nuances. For now I would like to preserve just this one case so that we can get testing and exploring the UX of potential "Bits" systems. In my opinion this code is no worse off than the existing code, so the fact that it doesn't solve more syntax irregularities seems like something that shouldn't be required.

**Comment by westonruter** [comment]
> This patch _only_ addresses a single kind of invalid comment leaving most other constructions alone.

Is there a list of all the possible invalid comment syntaxes?

**Comment by dmsnell** [comment]
> Is there a list of all the possible invalid comment syntaxes?

They are not listed separated like this but you can find them in the [parsing errors](https://html.spec.whatwg.org/#parse-errors) in the specification.

The Tag Processor draws its own classification which is retrievable with [`get_comment_type()`](https://developer.wordpress.org/reference/classes/wp_html_tag_processor/get_comment_type/), whereas it distinguishes:

 - Comments that are basically well-formed.
 - Comments that were abruptly closed.
 - Comments that appear as though someone wanted to write a CDATA section.
 - Comments that appear as though someone wanted to write an SGML processing instruction.
 - Comments that form from a variety of other invalid HTML constructions.

It separates out comments that appear as the result of being a closing tag with an invalid tag name. These it calls funky comments and these are the proposed syntax for Bits and other atomic shortcode like placeholders.

**Comment by dmsnell** [comment]
> How common are these other invalid comments?

@westonruter I took a list of the top-ranked domains and grabbed the first 296,000 pages at `/` for the domain.

[invalid-comments.txt](https://github.com/WordPress/wordpress-develop/files/15043143/invalid-comments.txt)

This file has ANSI terminal codes for clearer output. You can `cat` it and see colors…

Screenshot 2024-04-19 at 6 47 42 PM

File: src/wp-includes/kses.php (hunk 1 above)
Comment (dmsnell) [reply src/wp-includes/kses.php]: good call: I think this needs anchors. I'll review it and add them accordingly

File: src/wp-includes/kses.php (hunk 2 above)
Comment (ellatrix) [inline src/wp-includes/kses.php]: Stupid question: could we somehow reuse the same part of the regular expression `</[^a-zA-Z][^>]*>` below?

Comment (dmsnell) [reply src/wp-includes/kses.php]: Not stupid, but while we can do that, it's valuable IMO to leave this separate, as the _closing tag with an invalid tag name_ has an incredibly simple and straightforward syntax pattern: the tag name starts with _anything_ that isn't `a` - `z` (ASCII case-insensitive) and extends until the first `>`.

So by leaving this separate we can be even more careful about what we're examining. For example, opening tags with invalid tag names _are not interpreted as invalid comments_. The following pattern just isn't very robust in detecting normal tags let alone invalid ones, so as long as the correct approach is simpler being separate, I think it's worthwhile not joining them.

**Comment by ellatrix** [comment]
Thanks for omitting the others improvements, it really helps to understand that they are not related to allowing the some additional invalid comment syntaxes and are completely separate fixes. It also reduces the risk of a revert in the unlikely that case there's a problem with the additional fixes. Let's do those in a separate PR 👍

File: src/wp-includes/kses.php [hunk 3]
```diff
@@ -1080,12 +1092,48 @@ function _wp_kses_split_callback( $matches ) {
...
+	 * Preserve these comments and do not treat them like tags.
+	 */
+	if ( 1 === preg_match( '~^</[^a-zA-Z][^>]*>$~', $content ) ) {
+		$content     = substr( $content, 2, -1 );
+		$transformed = null;
+
+		while ( $transformed !== $content ) {
+			$transformed = wp_kses( $content, $allowed_html, $allowed_protocols );
+			$content     = $transformed;
+		}
```
Comment (dmsnell) [inline src/wp-includes/kses.php:1130]: this is copied from the existing comment behavior below

**Review by ellatrix — APPROVED** [review]
Looks good to me know, love the changes to make it more concise.

File: src/wp-includes/kses.php [hunk 4]
```diff
@@ -1080,12 +1092,48 @@ function _wp_kses_split_callback( $matches ) {
 function wp_kses_split2( $content, $allowed_html, $allowed_protocols ) {
 	$content = wp_kses_stripslashes( $content );
 
-	// It matched a ">" character.
+	/*
```
Comment (ellatrix) [inline src/wp-includes/kses.php:1097]: Should `wp_kses_split2` also get an updated `@since`?

**Review by ellatrix — COMMENTED** [review]
Should we maybe add an extra unit test case?

File: src/wp-includes/kses.php (hunk 4 above)
Comment (dmsnell) [reply src/wp-includes/kses.php:1097]: Yes, great idea. Thanks for catching it!

**Comment by dmsnell** [comment]
Found an issue with encoding HTML comments, I'll examine.

**Update**: Was testing locally on another branch and got confused; all is fine.

File: src/wp-includes/kses.php [hunk 5]
```diff
@@ -981,7 +982,18 @@ function wp_kses_split( $content, $allowed_html, $allowed_protocols ) {
 	$pass_allowed_html      = $allowed_html;
 	$pass_allowed_protocols = $allowed_protocols;
 
-	return preg_replace_callback( '%(<!--.*?(-->|$))|(<[^>]*(>|$)|>)%', '_wp_kses_split_callback', $content );
+	$token_pattern = <<<REGEX
+~
+	(                      # Detect comments of various flavors before attempting to find tags.
+		(<!--.*?(-->|$))   #  - Normative HTML comments.
+		|
+		</[^a-zA-Z][^>]*>  #  - Closing tags with invalid tag names.
```
Comment (westonruter) [inline src/wp-includes/kses.php:990]: These two lines are essentially the only change to the regex pattern (an addition).

**Review by westonruter — APPROVED** [review]
LGTM!

File: src/wp-includes/kses.php (hunk 3 above)
Comment (ellatrix) [reply src/wp-includes/kses.php:1130]: Could use an inline comment?

File: src/wp-includes/kses.php [hunk 6]
```diff
@@ -1069,23 +1081,61 @@ function _wp_kses_split_callback( $matches ) {
...
 	}
 
-	// Allow HTML comments.
+	/*
+	 * When a closing tag appears with a name that isn't a valid tag name,
+	 * it must be interpreted as an HTML comment. It extends until the
+	 * first `>` character after the initial opening `</`.
+	 *
+	 * Preserve these comments and do not treat them like tags.
+	 */
```
Comment (adamziel) [inline src/wp-includes/kses.php:1122]: Would the `>` character ever not terminate the string? In other words, what would change if you removed the $ from the regexp?

Comment (adamziel) [reply src/wp-includes/kses.php:1122]: Aha, I know now. Then we couldn’t assume bkw that the content starts at the at index two and ends at index minus one

File: src/wp-includes/kses.php [hunk 7]
```diff
@@ -1069,23 +1081,61 @@ function _wp_kses_split_callback( $matches ) {
...
+	if ( 1 === preg_match( '~^</[^a-zA-Z][^>]*>$~', $content ) ) {
```
Comment (adamziel) [inline src/wp-includes/kses.php:1123]: Is there any need to check for any invalid comment closers? I’m on a bus and can't remember exactly, but there were cases like triple dash or cdata-lookalike closer. Or does none of that matter with these tag-closer-like comments?

**Review by adamziel — APPROVED** [review]
I left a few notes I mean mostly as double-checking questions - I think you got everything right here and we’re good to go. Thank you! 

As a side note, I’d love to see a fuzzer trying all the different risky inputs with and without every kses PR

**Comment by dmsnell** [comment]
Merged in [[58418]](https://core.trac.wordpress.org/changeset/58418)
https://github.com/wordpress/wordpress-develop/commit/c756dfe8ca613d6c86d12fd1bd4ebf41cf60acff

**Pull request closed without merging**